keps/sig-network/3836-kube-proxy-improved-ingress-connectivity-reliability/README.md

```diff
@@ -0,0 +1,346 @@
+this, but....the Node object does not have a finalizer, so that wouldn't help.
+@thockin advocates for `.spec.unschedulable`, but I am inclined to say "that
+doesn't indicate if a Node is going away". `unschedulable` just says that no
+workloads can be assigned to the Node anymore. It is true that it's usually
+followed up by a drain...but I am not sure it's good to implement something
+which can easily kill ingress connectivity to a Node / cluster by simply doing
+`kubectl cordon node ...` (It would kill ingress connectivity to the entire
+cluster if the ingress controller is running on the node which is being cordoned)
+
+Maybe we need a KEP for a finalizer on the Node object (*troll face)
+<<[/UNRESOLVED]>>
+```
+
+```
+<<[UNRESOLVED HC-linked-service-fields ]>>
+Figure out which other cloud-providers configure the HC to target the service
+proxy for the HC
+<<[/UNRESOLVED]>>
+```
+
+## Motivation
+
+The motivation for each change is:
+
+1. Not including the `healthz` state can cause Kube-proxy to indicate to a load
+  balancer that it should send traffic to a Node simply because the endpoint is
+  scheduled there, even though Kube-proxy might not be healthy and successfully
+  managed to write the rules required for actually being able to forward traffic
+  to the endpoint.
+
+2. Workloads on terminating Nodes would benefit from connection draining as to
+  have their connections gracefully terminated. Workloads running on terminating
+  nodes get drained simultaneously with Kube-proxy today, this means that
+  there's no guarantee that connection draining will be enabled for them
+  currently. That might be the case if Kube-proxy is terminated before the
+  endpoint, but there's no guarantee of that. If Kube-proxy reacts to a "node is
+  about to go away" signal, then Kube-proxy can start prematurely failing the HC
+  and enable connection draining for these workloads.
+
+### Goals
+
+- Have kube-proxy accurately reflect its `healthz` state in its response to LB HCs
+  for class B) services.
+
+- Connection drain terminating Nodes
+
+### Non-Goals
+
+- Aligning cloud provider HCs for class A) services. Cloud providers like
+  Azure/AWS do not configure their HCs to point to the service proxy. Instead
+  they connect to the `NodePort` defined for the service. As to have them
+  benefit from the proposals of this KEP, they would need to change their
+  implementation.
+
+## Proposal
+
+#### Risk
+
+The risk with each proposal is:
+
+1. None. Class B) type services which benefitted from being scheduled on nodes
+   with a non-healthy Kube-proxy but which had successfully written the rules
+   required for traffic to reach the endpoint: will not benefit from that
+   anymore. But that's not deterministic...and that's not really a "risk".
+
+2. If the right field indicating "node is terminating / deleting" is
+   chosen/exists: None
+
+#### Mitigations
+
+None, since there are no Risks.
+
+## Design Details
```

Comment:
The HC server code is pretty brittle - I suspect some refactoring or retooling will be needed

Reply:
It's not too bad though. I (roughly) coded this KEP months ago, see: https://github.com/kubernetes/kubernetes/pull/111661/files